### PR TITLE
Ensure we emit a reference-equality check against null when generating .Equals

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/31283

### CODE REVIEW

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions_CreateEqualsMethod.cs [hunk 1]
```diff
@@ -309,6 +309,36 @@ public static IMethodSymbol CreateEqualsMethod(this Compilation compilation, Imm
             return statements.ToImmutableAndFree();
         }
 
+        private static void AddReferenceNotNullCheck(
+            SyntaxGenerator factory, Compilation compilation, ParseOptions parseOptions, SyntaxNode otherNameExpression, ArrayBuilder<SyntaxNode> expressions)
+        {
+            var nullLiteral = factory.NullLiteralExpression();
+            if (factory.SyntaxGeneratorInternal.SupportsPatterns(parseOptions))
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions_CreateEqualsMethod.cs]: 📝 Need to specifically check further for supporting _not_ patterns, and use `!(obj is null)` for C# 7

**Review by sharwell — CHANGES_REQUESTED** [review]

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions_CreateEqualsMethod.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions_CreateEqualsMethod.cs]: done.

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions_CreateEqualsMethod.cs [hunk 2]
```diff
@@ -177,9 +177,7 @@ public static IMethodSymbol CreateEqualsMethod(this Compilation compilation, Imm
 
                 // Ensure that the parameter we got was not null (which also ensures the 'as' test
                 // succeeded):
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions_CreateEqualsMethod.cs:179]:
```suggestion
                // Ensure that the parameter we got was not null (which also ensures the 'as' test succeeded):
```

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions_CreateEqualsMethod.cs [hunk 3]
```diff
@@ -309,6 +309,48 @@ public static IMethodSymbol CreateEqualsMethod(this Compilation compilation, Imm
...
+                            generator.ConstantPattern(nullLiteral))));
+            }
+            else if (generator.SupportsPatterns(parseOptions))
+            {
+                // if we support patterns then we can do `!(obj is null)`
+                expressions.Add(
+                    factory.LogicalNotExpression(
+                        generator.IsPatternExpression(otherNameExpression,
+                            generator.ConstantPattern(nullLiteral))));
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions_CreateEqualsMethod.cs:338]:
```suggestion
```

**Review by sharwell — APPROVED** [review]

**Pull request merged**